solr/modules/monitor/src/java/org/apache/solr/monitor/search/ReverseSearchComponent.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.handler.component.QueryComponent;
+import org.apache.solr.handler.component.ResponseBuilder;
+import org.apache.solr.handler.loader.JsonLoader;
+import org.apache.solr.monitor.AliasingPresearcher;
+import org.apache.solr.monitor.SolrMonitorQueryDecoder;
+import org.apache.solr.monitor.cache.MonitorQueryCache;
+import org.apache.solr.monitor.cache.SharedMonitorCache;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.schema.SchemaField;
+import org.apache.solr.update.DocumentBuilder;
+import org.apache.solr.util.SolrPluginUtils;
+import org.apache.solr.util.plugin.SolrCoreAware;
+
+public class ReverseSearchComponent extends QueryComponent implements SolrCoreAware {
+
+  public static final String COMPONENT_NAME = "reverseSearch";
+
+  private static final String SOLR_MONITOR_CACHE_NAME_KEY = "solrMonitorCacheName";
+  private static final String SOLR_MONITOR_CACHE_NAME_DEFAULT = "solrMonitorCache";
+  private String solrMonitorCacheName = SOLR_MONITOR_CACHE_NAME_DEFAULT;
+
+  private QueryDecomposer queryDecomposer;
+  private Presearcher presearcher;
+  private PresearcherFactory.PresearcherParameters presearcherParameters;
+
+  @Override
+  public void init(NamedList<?> args) {
+    super.init(args);
+    Object solrMonitorCacheName = args.remove(SOLR_MONITOR_CACHE_NAME_KEY);
+    if (solrMonitorCacheName != null) {
+      this.solrMonitorCacheName = (String) solrMonitorCacheName;
+    }
+    presearcherParameters = new PresearcherFactory.PresearcherParameters();
+    SolrPluginUtils.invokeSetters(presearcherParameters, args);
+  }
+
+  @Override
+  public void prepare(ResponseBuilder rb) throws IOException {
+    super.prepare(rb);
+    var req = rb.req;
+    var documentBatch = documentBatch(req);
+    var matcherSink =
+        new SyncSolrMatcherSink<>(
+            QueryMatch.SIMPLE_MATCHER::createMatcher,
+            new IndexSearcher(documentBatch.get()),
+            matchingQueries -> {
+              if (rb.isDebug()) {
+                rb.req
+                    .getContext()
+                    .put(
+                        ReverseSearchDebugComponent.ReverseSearchDebugInfo.KEY,
+                        new ReverseSearchDebugComponent.ReverseSearchDebugInfo(
+                            matchingQueries.getQueriesRun()));
+              }
+            });
+    Query preFilterQuery = presearcher.buildQuery(documentBatch.get(), getTermAcceptor(rb.req));
+    List<Query> mutableFilters =
+        Optional.ofNullable(rb.getFilters()).map(ArrayList::new).orElseGet(ArrayList::new);
+    rb.setQuery(new MatchAllDocsQuery());
+    mutableFilters.add(preFilterQuery);
+    var searcher = req.getSearcher();
+    MonitorQueryCache solrMonitorCache =
+        (SharedMonitorCache) searcher.getCache(this.solrMonitorCacheName);
+    SolrMonitorQueryDecoder queryDecoder = new SolrMonitorQueryDecoder(req.getCore());
```

Comment:
From code reading
```suggestion
    SolrMonitorQueryDecoder queryDecoder = new SolrMonitorQueryDecoder(this.queryDecomposer);
```

looks possible here, will attempt to do that.